vncviewer/GestureHandler.cxx

```diff
@@ -0,0 +1,180 @@
+/* Copyright 2019 Aaron Sowry for Cendio AB
+ * 
+ * This is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ * 
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ * 
+ * You should have received a copy of the GNU General Public License
+ * along with this software; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307,
+ * USA.
+ */
+
+#include <rfb/LogWriter.h>
+
+#include <X11/extensions/XI2.h>
+#include <X11/extensions/XInput2.h>
+
+#include "GestureHandler.h"
+
+static rfb::LogWriter vlog("GestureHandler");
+
+GestureHandler::GestureHandler() : state(GH_INITSTATE), tracked(), processed() {
+}
+
+int GestureHandler::registerEvent(void *ev) {
+#if !defined(WIN32) && !defined(__APPLE__)
+  XIDeviceEvent *devev = (XIDeviceEvent*)ev;
+
+  switch (devev->evtype) {
+    case XI_TouchBegin:
+      tracked.push_back(*devev);
+      break;
+
+    case XI_TouchUpdate:
+      // FIXME: We can distinguish between different two-touch gestures
+      // by looking at relative and absolute movement. This logic goes here.
+      // For now, just pass the update on to be processed normally.
+      processed.push_back({0,
+		           devev->event_x,
+		           devev->event_y,
+			   GH_TouchUpdate});
+      break;
+
+    case XI_TouchEnd:
+      // Does this event relate to a touch we're currently tracking?
+      bool matched = False;
+      for (int i = 0; i < tracked.size(); i++) {
+        if (tracked[i].detail == devev->detail) {
+          matched = True;
+	  break;
+	}
+      }
+
+      if (!matched)
+	// Related to either a XI_TouchBegin event we ignored earlier because
+	// we were already were already in a state, or a left-over touch from a
+	// gesture which has already ended. Just ignore it.
+        return 0;
+
+      if (!hasState(this->state))
+	// If we're not in a state already, then we're currently looking at
+	// a click. The currently tracked touches won't be part of a gesture,
+	// then - so let's expire them.
+        expireEvents();
+
+      // FIXME: By this point we should have a state, right?
+      //assert(hasState(this->state));
+
+      switch (this->state) {
+        case GH_LEFTBTN:
+	  processed.push_back({1,
+			       devev->event_x,
+			       devev->event_y,
```

Comment:
I think we might need some threshold here for clicks. I.e. you need to move your finger a certain distance, otherwise the press and release will both be at the same coordinate.

If we don't do this then you can get problems with clicks not registering as the system thinks it was a drag.

Reply:
Old code but I think I know what you mean and agree. I'm already working on this type of threshold for the scroll gesture (i.e. how far do fingers have to move before a right-click becomes a scroll?).